src/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/HostWorkspace/LanguageServerProjectLoader.cs

```diff
@@ -273,4 +370,74 @@ async Task LogDiagnosticsAsync(ImmutableArray<DiagnosticLogItem> diagnosticLogIt
             await toastErrorReporter.ReportErrorAsync(worstLspMessageKind, message, cancellationToken);
         }
     }
+
+    protected async ValueTask BeginLoadingProjectAsync(string projectPath, ProjectId primordialProjectId)
+    {
+        using (await _gate.DisposableWaitAsync(CancellationToken.None))
+        {
+            if (_loadedProjects.ContainsKey(projectPath))
+            {
+                return;
+            }
+
+            _loadedProjects.Add(projectPath, new ProjectLoadState.BeginLoadingWithPrimordial(primordialProjectId));
+            _projectsToLoadAndReload.AddWork(new ProjectToLoad(projectPath, ProjectGuid: null, ReportTelemetry: true));
+        }
+    }
+
+    protected async Task LoadProjectsAsync(ImmutableArray<(string ProjectPath, string? ProjectGuid)> projectPaths, CancellationToken cancellationToken)
+    {
+        using (await _gate.DisposableWaitAsync(cancellationToken))
+        {
+            foreach (var (path, guid) in projectPaths)
+            {
+                if (_loadedProjects.ContainsKey(path))
+                {
+                    continue;
+                }
+
+                _loadedProjects.Add(path, new ProjectLoadState.BeginLoading());
+                _projectsToLoadAndReload.AddWork(new ProjectToLoad(Path: path, ProjectGuid: guid, ReportTelemetry: true));
+            }
+
+            // TODO: since this method is used to implement 'OpenSolution' and 'OpenProjects',
+            // it seems like currently loaded projects which are not included in 'projectPaths' should be unloaded.
```

Comment:
OpenProjects, definitely not. I'd say in any case, this wouldn't be this method to decide -- that command would decide to unload stuff if needed, and maybe we'd have two commands of "open solution" (which closes the existing) or "add solution" which merges.

Just delete the comment.